masonry/src/app/event_loop_runner.rs

```diff
@@ -774,4 +774,65 @@ impl MasonryState<'_> {
             self.render_cx.set_present_mode(surface, present_mode);
         }
     }
+
+    /// Return references to the wgpu device and queue associated with current window.
+    /// TODO: multi-window interface will require more information to determine which
+    /// device and queue to return.
+    pub fn get_render_device_and_queue(&self) -> Option<(&wgpu::Device, &wgpu::Queue)> {
+        if let WindowState::Rendering { surface, .. } = &self.window {
+            let dev_id = surface.dev_id;
+            let device = &self.render_cx.devices[dev_id].device;
+            let queue = &self.render_cx.devices[dev_id].queue;
+            Some((device, queue))
+        } else {
+            None
+        }
+    }
+
+    /// Access to the wgpu adapter.
+    pub fn get_adapter(&self) -> Option<&wgpu::Adapter> {
+        if let WindowState::Rendering { surface, .. } = &self.window {
+            let dev_id = surface.dev_id;
+            Some(self.render_cx.devices[dev_id].adapter())
+        } else {
+            None
+        }
+    }
```

Comment:
These seem reasonable to me, in the "intentionally not very carefully reasoned about" way.